Catch EPIPE when connecting and retry

This patch will not be accepted by the upstream because the official supported
version of net-ssh is 3.0.2.

net-ssh 3.2 changed its behaviour when connecting.

Before:

* wait for server response until timout
* send local PROTO_VERSION

After:

* send local PROTO_VERSION (EPIPE can happen here if ProxyCommand is used, i.e.
  nc(1) and remote sshd is not running, yet)
* wait for server response

https://github.com/mfazekas/net-ssh/commit/ed2d4545e48ba591f4142192321a07c393e39849
--- plugins/communicators/ssh/communicator.rb.orig	2016-10-11 14:36:29.345475000 +0900
+++ plugins/communicators/ssh/communicator.rb	2016-10-11 14:40:51.506354000 +0900
@@ -414,6 +414,10 @@
         rescue Errno::ECONNREFUSED
           # This is raised if we failed to connect the max amount of times
           raise Vagrant::Errors::SSHConnectionRefused
+        rescue Errno::EPIPE
+          # This happens if net-ssh is 3.2.0 and nc is used in ProxyCommand
+          # and the VM sends RST TCP packets
+          raise Vagrant::Errors::SSHConnectionRefused
         rescue Errno::ECONNRESET
           # This is raised if we failed to connect the max number of times
           # due to an ECONNRESET.
